### ISSUE TITLE
Not Receiving Data

### ISSUE THREAD

**Issue by jeffcoh23** [eredden/Veeder-Root-TLS-Socket-Library#2]
Hi,

Thanks for making this library!  I am trying to reverse engineer communicating with a TLS-350/450 as a project for a friend (I know very little about these devices) and have converted a lot of your code to Ruby (my preferred language).

I am getting caught on this line in your `tls_socket.py` file on line `chunk = socket.recv(data_size)` using `response = socket.execute("i10100")`.  The TLS seems to wait a good bit of time ~30 seconds and then sends back `nil`.  Any ideas why this is?  If it helps, when I execute `response = socket.execute("random_command")` it sends back the proper error code, so I know I'm communicating with the device.

Any insight is appreciated!

**Comment by eredden** [comment]
Hello Jeff! I originally had the `execute` function take in a timeout argument for how long it could wait prior to trying to receive the data. I changed this very recently to instead try receiving data every second for 30 seconds until the end of transmission character (also known as ETX, it is ASCII code 003) is found.

My best guess is that your program is not ending the loop once the ETX is found. Try including a print statement inside the loop and see if you're receiving the chunks of data. If you are receiving everything but the loop continues to go on, then this is likely your issue.

If you happen to have a repository containing that Ruby code, I would love to take a look at it.

**Comment by jeffcoh23** [comment]
Thanks for the quick response @eredden! I've put in debuggers via `binding.pry` (line 40 on `tls_connect.rb` in the zip file) and when I checked the `chunk` value it always gave `nil` unless I gave some nonsensical command like `your_command` into the `execute` function, in which case, I would get `Invalid Command`.  My hunch is either:

1. There is something on the ATG blocking the response - possibly a security config I'm unaware of?  As I said, I'm not super familiar with these devices.
2. I messed up converting your script and the wrong `execute` value is being sent to the ATG and consequently not being read correctly.  

Regarding the files, I put them in a zip below.  For simplicity, I just have it run like a script instead of a CLI program.  The script executes by running `ruby execute.rb`.  If you're familiar with Ruby it should be pretty straightforward, but if its not clear, I'd be happy to explain further.  

If I had more time this afternoon I would pull down your repo and see if the same issue happens with the TLSs I have access to.  Tomorrow I can give it a go. 

[tls_files.zip](https://github.com/eredden/Veeder-Root-TLS-Socket-Library/files/14968493/tls_files.zip)

**Comment by jeffcoh23** [comment]
Edit: I did have time to try it out.  Makes me now think the issue is with the ATG setup even though I am able to connect without any issues... 🤔  

```
➜  Veeder-Root-TLS-Socket-Library git:(main) ✗ python3 src/tls_client.py <"IP_ADDRESS"> <PORT>
You are connected to <IP_ADDRESS> using port <PORT>.

>> I10100
Transmission failed.
```

**Comment by eredden** [comment]
There's a security feature on the automatic tank gauges that would prevent a command from being sent without a 6-digit security code being prepended to the command. Could it be possible that this feature has configured on that ATG?

Another thing could be that the start of header was adjusted for that machine. If that's the case, you would need to modify `tls_client.py` to set the start of header argument in the `execute()` function.

Let me know if you happen to figure out anything about this. I'll try this on the ATGs I have access to and see if I can replicate this error on those.

**Comment by eredden** [comment]
Tested this out on three different TLS automatic tank gauges in varying locations across the globe by running `I10100`, and things do appear to be working OK on my end for all of them. There's a fair chance this could be specific to the automatic tank gauge you are working with. Definitely check with your friend if the start of header or security code for the automatic tank gauge has been adjusted for security purposes.

If either those happen to be the case, read my previous message for recommendations. Let me know what you find!

**Comment by jeffcoh23** [comment]
Awesome, thank you @eredden - I'll let you know!  Just curious, is changing the start of the header a security feature as well?

**Comment by eredden** [comment]
Not quite sure why someone would want to change the header -- could be security, unsure. 

**EDIT:** I just now saw your previous message which attached the .zip file for your Ruby code. Could it be possible that Ruby is mistakenly interpreting your start of header as a string rather than a bytestring? This line of code converts your command into a UTF-8 encoded bytestring, but I am unsure as to whether or not the same is done to the `soh` variable.
```ruby
byte_command = soh + command.encode("utf-8")
```

If there is a way to check types of variables in Ruby, determine whether or not `soh` is being turned into a bytestring or not. That could definitely cause you issues if that is not the case.

**Comment by jeffcoh23** [comment]
> Not quite sure why someone would want to change the header -- could be security, unsure.
> 
> **EDIT:** I just now saw your previous message which attached the .zip file for your Ruby code. Could it be possible that Ruby is mistakenly interpreting your start of header as a string rather than a bytestring? This line of code converts your command into a UTF-8 encoded bytestring, but I am unsure as to whether or not the same is done to the `soh` variable.
> 
> ```ruby
> byte_command = soh + command.encode("utf-8")
> ```
> 
> If there is a way to check types of variables in Ruby, determine whether or not `soh` is being turned into a bytestring or not. That could definitely cause you issues if that is not the case.

@eredden Given that your python script didnt work for my ATGs I'm thinking there must be one of those aforementioned reasons (header or security code) that is the issue.  I'm going to track that down in the next few days and will let you know.  Your help has been much appreciated!

*Edit:* If I do need to add a security code, is there a way in your `tls_socket.py` file to add it in and if not, do you know where it would be appended? 

*Extra Edit*: I logged into the content management solution they use to look at the configs and I saw the `SID #` is blank - I think this may be the security code?

**Comment by eredden** [comment]
"Given that your python script didnt work for my ATGs I'm thinking there must be one of those aforementioned reasons (header or security code) that is the issue."

That makes sense, not sure why that did not cross my mind! As for if there is a way to do it with `tls_socket.py`, there is not. There's a certain command you would need to run to do that, so you would probably need to establish a raw socket connection to that automatic tank gauge through something like PuTTY and run the command there.

As for that `SID #` field, I am not sure what that is associated with. I would think that the start of header would be labelled `SOH` instead as the manuals for these automatic tank gauges always abbreviate it like that.

**Comment by eredden** [comment]
I think I have mixed up the start of header and end of transmission characters. I have re-read the Veeder-Root manual in my repository and it looks like it's actually the end of transmission (ETX) character that can be adjusted using command `S53100f`. The security code can be adjusted using command `S53600f`.

My apologies for the mix up! Unfortunately, I have not yet encountered a way to see the current value of either of these two fields.

![code-change](https://github.com/eredden/Veeder-Root-TLS-Socket-Library/assets/155211750/42705b13-8b51-4337-9eb7-75c638baf550)
![etx-change](https://github.com/eredden/Veeder-Root-TLS-Socket-Library/assets/155211750/d92d1f04-c4c9-4a24-af3b-6558b140d922)

**Comment by GrantBlack** [comment]
I work on ATGs a fair bit and have couple in a test lab.

@jeffcoh23 - Whenever you test code for ATGs, start with just using telnet. 
You can test commands that way very easily; just Windows telnet <ip> <port> then CTRL-A, I20100 or similar command.
Note capital I for ASCII response, lowercase i for binary

We never switch on security for serial commands as can be a pain in the field and adds very little to security. But before looking at code, test manually with telnet.  If a TLS450, you can also use HTTPS to check settings. Veeder Root Inform app is another option but it just automates the sending/receiving of serial commands

**Comment by jeffcoh23** [comment]
Thank you both @GrantBlack and @eredden for the helpful debugging info.  Unfortunately, my friend gave me some incorrect info and it turns out there are security codes on nearly every ATG they have!  I found a few that don't have them and this repo and my Ruby adaptation worked flawlessly.

Now I have to modify my script to make it work with the security codes.  @eredden I've reread your posts above and I find their documentation somewhat confusing.  As an excuse, in my day to day work I never interface with devices so this has been a confusing, but fun learning experience.  Is your message suggesting I just add `S53600XXXXXX\x03` to the etx?  In the proposed code below, is doing something like this where the `XXXXXX` is the 6 character alphanumeric security code on the right track?

```
 def execute(self, command: str, etx: bytes = b"S53600XXXXXX\x03") -> str:
```

**Edit:** I just tapped into one of the facilities to get tank inventory readings (command: `I20100`), but obviously it either comes back in display or computer format.  Is there a simpler way people typically parse this data to put it into database form (e.g. PostgreSQL) that I'm unaware of?

**Comment by eredden** [comment]
I mention using security codes in my README file but I did not provide an exact example. See below excerpts:

"If your TLS system needs security codes in front of the commands, then you would simply type the security code before the command" is what I stated in my documentation, e.g. `XXXXXXi10100`. It would not go after the command like in your example. This is also described in the Veeder-Root manual in the `docs` folder of my repository.
![message-format](https://github.com/eredden/Veeder-Root-TLS-Socket-Library/assets/155211750/27494f2b-a0d3-4590-8f96-5bf4c08f9151)

As for parsing the responses from the Veeder-Root ATGs, I have some functions in `tls_3xx_functions.py` that do this for some of the commands. There is no particularly easy way to do this as the responses for these commands vary heavily.

Let me know if there is anything else you would like help with.
